### PR TITLE
Correct max iteration warnings in AirloopHVAC:UnitarySystem

### PR DESCRIPTION
**Pull request by rraustad**
## Pull request overview

Example file HVACTemplate:UnitarySystem showed max iteration warnings on design day simulations (#5531). Corrected by removing the "set" of available flow rate in the parent object (AirloopHVAC:UnitarySystem) and allowing the air loop to control convergence. Also added a test to see if the system is oscillating on operating mode (e.g., cooling then heating then cooling) in the same iteration, and if so, control to the zone load instead of the system output. 
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [ ] If defect, results of running current develop vs this branch should exhibit the fix
- [ ] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [ ] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] ExpandObjects changes??
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

…llow a wider range of flow rates for cycling fan systems.

### CODE REVIEW

**Comment by rraustad** [comment]
@Myoldmopar @mjwitte @EnergyArchmage simple changes so far.

**Comment by mjwitte** [comment]
@rraustad Question - when unitary systems change speed or change flow between heating/cooling/no-load, do they touch maxavail? If they are, maybe that's part of the problem.

**Comment by rraustad** [comment]
Yes it does, but only because I didn't know what was happening.

```
if ( GetCurrentScheduleValue( UnitarySystem( UnitarySysNum ).SysAvailSchedPtr ) > 0.0 && ( ( FanOn || TurnFansOn ) && ! TurnFansOff ) ) {
    if ( UnitarySystem( UnitarySysNum ).ControlType == SetPointBased ) {
        // set point based equipment should use VAV terminal units to set the flow.
        // zone equipment needs to set flow since no other device regulates flow (ZoneHVAC /= AirLoopEquipment)
        if ( !UnitarySystem( UnitarySysNum ).AirLoopEquipment ) {
            Node( InletNode ).MassFlowRate = AverageUnitMassFlow;
            Node( InletNode ).MassFlowRateMaxAvail = AverageUnitMassFlow;
        }
        if ( AverageUnitMassFlow > 0.0 ) {
            OnOffAirFlowRatio = 1.0;
        } else {
            OnOffAirFlowRatio = 0.0;
        }
    } else {
        Node( InletNode ).MassFlowRate = AverageUnitMassFlow;
        Node( InletNode ).MassFlowRateMaxAvail = AverageUnitMassFlow;
        if ( AverageUnitMassFlow > 0.0 ) {
            OnOffAirFlowRatio = CompOnMassFlow / AverageUnitMassFlow;
        } else {
            OnOffAirFlowRatio = 0.0;
        }
    }
```

**Comment by mjwitte** [comment]
What about the other unitary furnace and heat pump systems?

**Comment by rraustad** [comment]
I see similar in Furnaces and WSHPs. I have a feeling that these might be able to be deleted once we figure all this out. I think it should only be the child components that do anything with MaxAvail.

**Comment by rraustad** [comment]
VSHeatPumpWaterToAirEquationFit (#5447) now runs without max iteration warnings in Denver or Chicago. However, other example files now exhibit max iteration warnings which were not present previously.

**Comment by mjwitte** [comment]
That's the kind of problem (fix some, break some) that I ran into with #5159

**Comment by rraustad** [comment]
This is not working for AirTerminal:SingleDuct:ConstantVolume:Reheat because they want to operate at the MaxAvail instead of calculating a flow. I guess they wouldn't be constant volume if they calculated a flow rate?

```
    QToHeatSetPt = ZoneSysEnergyDemand( ZoneNum ).RemainingOutputReqToHeatSP; // The calculated load from the Heat Balance
    MassFlow = SysInlet( SysNum ).AirMassFlowRateMaxAvail; // System massflow is set to the Available
```

**Comment by mjwitte** [comment]
That may need to change.  What does AirTerminal:SingleDuct:ConstantVolume:Uncontrolled do?

**Comment by rraustad** [comment]
The direct air unit, AirTerminal:SingleDuct:Uncontrolled, operates at the inlet flow. I guess this should also work for the AirTerminal:SingleDuct:ConstantVolume:Reheat object.

```
MassFlowRate = Node( DirectAir( DirectAirNum ).ZoneSupplyAirNode ).MassFlowRate;
```

**Comment by mjwitte** [comment]
Maybe, but there's probably also some logic in Uncontrolled that starts with a request for the terminal unit max flow.  And that may be missing in Reheat.

**Comment by rraustad** [comment]
This doesn't look good. The TU boxes were being limited in their flow by the same code I just changed. The air loop wants to operate at some fraction of it's design flow, minus some other garbage (below). I for the life of me can't figure out how in the world we could converge with everyone doing their own thing. Now I see how we got to a point where Min/MaxAvail became a lock on flow. And the splitter determines it's inlet by summing the outlets, which are set by the TUs. If the splitter inlet does not match the air loop outlet, round and round we go.

The warmest temp SP manager or NCycVent manager can set ReqSupplyFrac lower than 1. Some of the terminal units operate at the max avail. Like I said before, the change made might work for cycling systems but not VAV. We've selectively coded ourselves into a corner.

I would have never thought I'd see this on the air loop for VAV systems (i.e., it should just operate at whatever flow is on the inlet). I could see where some equipment types would operate at their max flow \* ReqSupplyFrac, but the other garbage here blows that up. When these other terms are non-zero, the zone equipment better know what to do. The Min/MaxAvail has been a surrogate to choke the flow because of things like this.

```
Node( NodeNumIn ).MassFlowRate = AirLoopFlow( AirLoopNum ).DesSupply * AirLoopFlow( AirLoopNum ).ReqSupplyFrac - ( AirLoopFlow( AirLoopNum ).ZoneExhaust - AirLoopFlow( AirLoopNum ).ZoneExhaustBalanced - AirLoopFlow( AirLoopNum ).RetFlowAdjustment );
```

The good news is that there are just a handful of files that have issues with flow. In a lot of cases, max iterations were eliminated. In a few, they were added. Files with warmest temp manager, the constant volume reheat box, I'll bet night cycle vent, and probably some others I haven't identified yet.

**Comment by mjwitte** [comment]
So, what if you undo these code changes and change the unitary system to stop touching maxflowavail?

**Comment by rraustad** [comment]
This change worked for the defect file, but I am worried it won't work for 2 UnitarySystems on the main branch.

**Comment by rraustad** [comment]
Documenting what I've seen so far.
With the change in UnitarySystem (2 lines commented MaxAvail), 
UnitarySystem_VSHeatPumpWaterToAirEquationFit.idf - if 1622 in HVACManager is commented, no max iter, if not commented, max iter (tested to see if I should have also included 1622 with new checkin)
UnitarySystem_FurnaceWithDXSystemRHcontrol.idf - shows big diff's but really no difference. Heating coil turns on at different times but with same magnitude. T and RH in control zone look nearly identical. Coil heating rate and cool coil sensible energy profiles look similar to before (except for heat coil timing).

maxiter_unitarysystemfurnacerhcontrol

**Comment by rraustad** [comment]
Unit tests now need to set inlet node MaxAvail otherwise the fan won't turn on. One line in each of 3 units tests and they all pass now.

**Comment by rraustad** [comment]
Documentation of results (also see above):
5ZoneWLHPPlantLoopTower - has 2 new max iteration warnings (none before)

5ZoneWarmestVFD_FCMAuto - no new errors, AirTerminal:SingleDuct:VAV:Reheat, results look nearly identical.

5zonewarmest_maxiter

5ZoneWaterLoopHeatPump - no new errors, AirTerminal:SingleDuct:Uncontrolled, results look similar but sim converged to different point and there are some big diff's based on a different operating point (e.g., zone temps are 0.01 different and heating rates are 10s of watts different).

5zonewlhp_maxiter

**Comment by mjwitte** [comment]
@rraustad Note that 5ZoneWLHPPlantLoopTower.idf and 5ZoneWaterLoopHeatPump.idf have issues with a mix of autosized and hard sized flow rates.  Fixes are pending over at #5159

**Comment by rraustad** [comment]
CoolingCoilFreezeProtection is a new example file - as with other diff files, convergence is now different. Space temp is maintained, air flow and coil capacity vary (cooling) as expected.

HVACTemplate-5ZoneUnitarySystem - max iterations are eliminated.
HeatPumpCycFanWithEcono - 8 max iter warnings were added.
UnitarySystem_5ZoneWaterLoopHeatPump - coil control failed warnings added, max iter warnings added.
UnitarySystem_HeatPumpAuto - no new warnings, looks same as before.
unitarysystem_heatpumpauto_maxiter

UnitarySystem_MultiSpeedCoils_SingleMode - new example file, no new warnings, big diff's.

UnitarySystem_VSHeatPumpWaterToAirEquationFit - max iter eliminated, big diff's.
unitarysys_vsheatpumpwatertoairequafit_maxiter

UnitarySystem_WaterCoils_wMultiSpeedFan - no new warnings, big diff's.
unitarysys_watercoils_wmultispeedfan_maxiter

VSHeatPumpWaterToAirEquationFit - max iter eliminated, big diff's.
vsheatpumpwatertoairequationfit_maxiter

VSHeatPumpWaterToAirWithRHControl - max iter eliminated, Convergence warning, big diff's.
vsheatpumpwrhcontrol_maxiter

**Comment by rraustad** [comment]
You can see from these figures that the simulations have not really changed fundamentally. Yes, there are differences, but there are also differences in convergence with eliminating the resetting of MaxAvail every few iterations. I'm comfortable with the differences I see. Any comments?

File: tst/EnergyPlus/unit/HVACUnitarySystem.unit.cc [hunk 1]
```diff
@@ -1036,6 +1036,7 @@ TEST_F( EnergyPlusFixture, UnitarySystem_GetInput ) {
 	Schedule( 1 ).CurrentValue = 1.0;
 	DataGlobals::BeginEnvrnFlag = true;
 	DataEnvironment::StdRhoAir = PsyRhoAirFnPbTdbW( 101325.0, 20.0, 0.0 ); // initialize RhoAir
+	Node( InletNode ).MassFlowRateMaxAvail = UnitarySystem( 1 ).MaxCoolAirVolFlow * StdRhoAir;
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/HVACUnitarySystem.unit.cc:1058]: @rraustad What if the heating or no load flow rate is greater than MaxCoolAirVolFlow?

Comment (rraustad) [inline tst/EnergyPlus/unit/HVACUnitarySystem.unit.cc:1058]: Well then you would use the max flow rate. These are unit tests where heat/cool flow are equal. Are you saying I should have used max(UnitarySystem( 1 ).MaxCoolAirVolFlow, UnitarySystem( 1 ).MaxHeatAirVolFlow, ...) \* StdRhoAir here instead for maintenance sake if someone copied one of these unit tests?

Comment (mjwitte) [inline tst/EnergyPlus/unit/HVACUnitarySystem.unit.cc:1058]: Duh - never mind - didn't notice this was in the unit test.

**Comment by rraustad** [comment]
The figures above were a comparison of V8.4 with this branch. To pick up other example file comparisons, I used current develop to compare with this branch to include new example files. All file appear to operate the same now as before, except the CoolingCoilFreezingPrevention file which is more well behaved now.

5zonewlhpplantlooptower_maxiter

coolingcoilfreezingprevention_maxiter

heatpumpcycfanwithecono_maxiter

unitarysys_5zonewaterloopheatpump_maxiter

unitarysys_multispeedcoils_singlemode_maxiter

**Comment by mjwitte** [comment]
@rraustad I merged in develop and resolved the small conflict.  These changes do eliminate max iteration warnings, but these new warnings seem to indicate that something is still out of synch between the system and the coil:

```
   *************  ** Warning ** Coil:Heating:DX:SingleSpeed "SYS 3 HEAT PUMP AIRSOURCE SNGLSPD HEATING COIL" - Air volume flow rate per watt of rated total heating capacity is out of range error continues...
   *************  **   ~~~   **   This error occurred 492 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=1.656543E-005  Min=2.361783E-007
   *************
   *************  ** Warning ** Temperature out of range [-100. to 200.] (PsyPsatFnTemp)
   *************  **   ~~~   **   This error occurred 75 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=2033.582031 C  Min=204.997078 C
   *************
```

Note the max temp of 2033C being sent to the psych routines.  Added two flavors of this file to defect files, one with design days plus 1 week of Jan, Apr, and Jul.  And another for annual.  The annual run hits 4000+C.

**Comment by rraustad** [comment]
Well, here it is. The unitary system puts 0.34 kg/s on the inlet node to see what the system max (HeatPLR=1) output is 

```
CalcUnitarySystemToLoad( UnitarySysNum, AirLoopNum, FirstHVACIteration, CoolPLR, HeatPLR, OnOffAirFlowRatio, SensOutputOn, LatOutputOn, HXUnitOn, _, _, CompressorONFlag );
FullSensibleOutput = SensOutputOn;
```

and the fan is simulated first.

```
Fan( FanNum ).MassFlowRateMaxAvail = min( Node( OutletNode ).MassFlowRateMax, Node( InletNode ).MassFlowRateMaxAvail );
```

Node( OutletNode ).MassFlowRateMax = 0.3403 kg/s.
Node(InletNode).MassFlowRateMaxAvail = 0.102 kg/s.
Node( InletNode ).MassFlowRate = 0.3403 kg/s.

Then the fan chokes the flow.

```
Fan( FanNum ).InletAirMassFlowRate = min( Node( InletNode ).MassFlowRate, Fan( FanNum ).MassFlowRateMaxAvail );
```

At this time step, it took 5 iterations before this happened (I had to hit F5  5 times).

This mass flow rate is passed to the coil with PLR = 1 and the warning is tripped.
I don't have to explain much more than that after we've seen what we've seen.

**Comment by mjwitte** [comment]
Soooo, why should the fan care about it's inlet node MassFlowRateMaxAvail?

**Comment by mjwitte** [comment]
And so, maybe that's why the unitarysystem was resetting MassFlowRateMaxAvail.  For the case where the fan is a child of the unitary system, it seems that the unitary system should be dictating the flow rate, especially in a multispeed/variablespeed situation.

File: src/EnergyPlus/HVACUnitarySystem.cc [hunk 2]
```diff
@@ -10777,7 +10777,7 @@ namespace HVACUnitarySystem {
 				// zone equipment needs to set flow since no other device regulates flow (ZoneHVAC /= AirLoopEquipment)
 				if ( !UnitarySystem( UnitarySysNum ).AirLoopEquipment ) {
 					Node( InletNode ).MassFlowRate = AverageUnitMassFlow;
-					Node( InletNode ).MassFlowRateMaxAvail = AverageUnitMassFlow;
+//					Node( InletNode ).MassFlowRateMaxAvail = AverageUnitMassFlow; #5531 max iterations
```
Comment (mjwitte) [inline src/EnergyPlus/HVACUnitarySystem.cc]: Do we have a test file with UnitarySystem as zone equipment?  This line may need to stay.

**Comment by rraustad** [comment]
Because the throttling of MaxAvail is telling the fan not to exceed those limits. I think the code that allows the VAV systems to converge should be moved to the respective TU's. I also think each TU should determine the flow rate it should operate at. Either by reading the inlet node mass flow or by calculating. IF AirTerminal:SingleDuct:ConstantVolume:Reheat wants or needs to operate at MaxAvail then that specific TU must determine it's limits.

This code is fairly straight-forward. It calculates a flow ratio on the air loop supply outlet node. Then applies that ratio to the TU MaxAvail. We can open up the flow limits for direct air units by moving this code to specific TUs.

```
SupplyNode = AirToZoneNodeInfo( AirLoopIndex ).AirLoopSupplyNodeNum( SupplyIndex );
if ( Node( SupplyNode ).MassFlowRate > 0.0 ) {
    if ( ( Node( SupplyNode ).MassFlowRateSetPoint - Node( SupplyNode ).MassFlowRate ) > HVACFlowRateToler * 0.01 ) {
      FlowRatio = Node( SupplyNode ).MassFlowRate / Node( SupplyNode ).MassFlowRateSetPoint;
      for ( ZonesCooledIndex = 1; ZonesCooledIndex <= AirToZoneNodeInfo( AirLoopIndex ).NumZonesCooled; ++ZonesCooledIndex ) {
        TermInletNode = AirToZoneNodeInfo( AirLoopIndex ).TermUnitCoolInletNodes( ZonesCooledIndex );
        Node( TermInletNode ).MassFlowRateMaxAvail = Node( TermInletNode ).MassFlowRate * FlowRatio;
      }
```

This line would be moved to Init in each TU. I'm not sure how the air loop MaxAvail is getting passed back to the inlet of the air loop.

```
        Node( TermInletNode ).MassFlowRateMaxAvail = Node( TermInletNode ).MassFlowRate * FlowRatio;
```

**Comment by rraustad** [comment]
Geeez, zone equipment. Yes, if zone equipment then someone has to set MaxAvail.

**Comment by rraustad** [comment]
OK to check in my UnitarySystem file that has OA, air loop, and zone equipment in one file as:

_UnitarySystem_Zone_Airloop_OASystem.idf 

of maybe this would benefit users?

**Comment by mjwitte** [comment]
Regarding the fan and maxavail, the terminal units already have a say in the mass flow rates.  That's why the [zonesplitters](https://github.com/NREL/EnergyPlus/blob/da43757921bbf1606cd38230b5972b4b8396d692/src/EnergyPlus/SplitterComponent.cc#L530-L537) and [plenums](https://github.com/NREL/EnergyPlus/blob/a070fda4208c2abc79ab64729a4197d30d313cb8/src/EnergyPlus/ZonePlenum.cc#L958-L965) get called twice - once to pass things forward, and a second time to pass flow requests backward to the loop.  After this release, we need to step back and re-evaluate all the various flow resolution steps in air loops.  For now, just trying to make some incremental improvements.

**Comment by mjwitte** [comment]
_UnitarySystem_Zone_Airloop_OASystem.idf seems complicated.  Would be useful to have an example file that just has unitary system as zone equipment, no air loops, no other zone equipment.

**Comment by rraustad** [comment]
@Myoldmopar the i386-Windows-7-VisualStudio-12 CI has issues.

**Comment by Myoldmopar** [comment]
Weird.  Doesn't _appear_ like a local computer error, maybe a network interruption.  I'm going to click on the Unhandled Failure link, and at the top of the page, there is a link to delete the test results and force a rebuild.  I'll do that.  Hopefully the fluke will have passed.

**Comment by mjwitte** [comment]
@rraustad This is causing new warnings in [UnitarySystem_5ZoneWaterLoopHeatPump](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2016-03/a070fda4208c2abc79ab64729a4197d30d313cb8-8f9b4ae4a59138e2da6ccbd1758ee4262e18a69a/UnitarySystem_5ZoneWaterLoopHeatPump/x86_64-MacOS-10.9-clang/eplusout.err.diff.html).
It eliminates max iteration warnings in HVACTemplate-5ZoneUnitarySystem-Annual but creates temperature issues discussed above.  Not sure we're there yet.

**Comment by rraustad** [comment]
UnitarySystem_5ZoneWaterLoopHeatPump:

```
Program Version,EnergyPlus, Version 8.5.0-edc59f04f3, YMD=2016.03.18 18:29,IDD_Version 8.5.0
 ************* Testing Individual Branch Integrity
 ************* All Branches passed integrity testing
 ************* Testing Individual Supply Air Path Integrity
 ************* All Supply Air Paths passed integrity testing
 ************* Testing Individual Return Air Path Integrity
 ************* All Return Air Paths passed integrity testing
 ************* No node connection errors were found.
 ************* Beginning Simulation
 ************* Simulation Error Summary *************
 ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
 ************* EnergyPlus Sizing Error Summary. During Sizing: 0 Warning; 0 Severe Errors.
 ************* EnergyPlus Completed Successfully-- 0 Warning; 0 Severe Errors; Elapsed Time=00hr 01min  8.42sec
```

File: src/EnergyPlus/HVACUnitarySystem.cc [hunk 3]
```diff
@@ -1689,11 +1704,25 @@ namespace HVACUnitarySystem {
...
+					ZoneLoad = QToCoolSetPt;
+				} else if( QToHeatSetPt > 0.0 ) {
+					HeatingLoad = true;
+					CoolingLoad = false;
+					ZoneLoad = QToHeatSetPt;
+				} else {
+					HeatingLoad = false;
+					CoolingLoad = false;
+					ZoneLoad = 0.0;
+				}
```
Comment (rraustad) [inline src/EnergyPlus/HVACUnitarySystem.cc:1746]: This model should meet the load 1st time, if > 4 iterations, lock down mode.

File: src/EnergyPlus/HVACUnitarySystem.cc [hunk 4]
```diff
@@ -922,6 +936,7 @@ namespace HVACUnitarySystem {
 
 			} // from IF(UnitarySystem(UnitarySysNum)%SuppHeatCoilType_Num == Coil_HeatingSteam) THEN
 		} // from IF( FirstHVACIteration ) THEN
+		iterationCounter += 1;
```
Comment (rraustad) [inline src/EnergyPlus/HVACUnitarySystem.cc]: Count iterations within the model.

File: src/EnergyPlus/HVACUnitarySystem.cc [hunk 5]
```diff
@@ -10786,7 +10815,9 @@ namespace HVACUnitarySystem {
 				}
 			} else {
 				Node( InletNode ).MassFlowRate = AverageUnitMassFlow;
-				Node( InletNode ).MassFlowRateMaxAvail = AverageUnitMassFlow;
+				if ( !UnitarySystem( UnitarySysNum ).AirLoopEquipment ) {
+					Node( InletNode ).MassFlowRateMaxAvail = AverageUnitMassFlow; // #5531 zone equipment needs MaxAvail set or fan will not turn ON
+				}
```
Comment (rraustad) [inline src/EnergyPlus/HVACUnitarySystem.cc:10842]: zone equipment specific...must set MaxAvail

File: src/EnergyPlus/HVACUnitarySystem.cc [hunk 6]
```diff
@@ -472,8 +479,12 @@ namespace HVACUnitarySystem {
...
 			}
 
-			if ( ( HeatActive ) && ( AirLoopControlInfo( AirLoopNum ).CanLockoutEconoWithCompressor || AirLoopControlInfo( AirLoopNum ).CanLockoutEconoWithHeating ) ) {
-				AirLoopControlInfo( AirLoopNum ).ReqstEconoLockoutWithHeating = true;
+			if ( present (HeatActive) ) {
+				if ( ( HeatActive ) && ( AirLoopControlInfo( AirLoopNum ).CanLockoutEconoWithCompressor || AirLoopControlInfo( AirLoopNum ).CanLockoutEconoWithHeating ) ) {
+					AirLoopControlInfo( AirLoopNum ).ReqstEconoLockoutWithHeating = true;
+				} else {
+					AirLoopControlInfo( AirLoopNum ).ReqstEconoLockoutWithHeating = false;
+				}
```
Comment (rraustad) [inline src/EnergyPlus/HVACUnitarySystem.cc:487]: result of zone, air loop, OA sys example file test. "HeatActive" flag wasn't passed in from MixedAir.

File: tst/EnergyPlus/unit/HVACUnitarySystem.unit.cc [hunk 7]
```diff
@@ -4020,6 +4022,7 @@ TEST_F( EnergyPlusFixture, UnitarySystem_MultiSpeedCoils_SingleMode ) {
 
 	int Iter;
 	Real64 StdRhoAir = 1.2;
+	Node( InletNode ).MassFlowRateMaxAvail = UnitarySystem( UnitarySysNum ).MaxCoolAirVolFlow * StdRhoAir;
 	Iter = 4;
 	UnitarySystem( UnitarySysNum ).CoolVolumeFlowRate( Iter ) = UnitarySystem( UnitarySysNum ).MaxCoolAirVolFlow * DesignSpecMSHP( 1 ).CoolingVolFlowRatio( Iter );
 	UnitarySystem( UnitarySysNum ).CoolMassFlowRate( Iter ) = UnitarySystem( UnitarySysNum ).CoolVolumeFlowRate( Iter ) * StdRhoAir;
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HVACUnitarySystem.unit.cc:4028]: unit tests must set MaxAvail to turn the fan on

File: src/EnergyPlus/HVACUnitarySystem.cc (hunk 3 above)
Comment (mjwitte) [inline src/EnergyPlus/HVACUnitarySystem.cc:1746]: @rraustad I'm a little uncomfortable with arbitrarily choosing 4 iterations as the time to lock down.  Handing this off to @Myoldmopar and @EnergyArchmage to decide if this should go in 8.5 or if it needs more thought.

**Comment by rraustad** [comment]
Well this actually looks better now. This is HVACTemplate-5ZoneUnitarySystem modified to run design days. Max iter is now gone and zone temps look better now.

hvactemplate-5zoneunitarysystem_designdays

File: src/EnergyPlus/HVACUnitarySystem.cc (hunk 3 above)
Comment (rraustad) [inline src/EnergyPlus/HVACUnitarySystem.cc:1746]: I could pick a value closer to 20, but that would mean the sim would unnecessarily iterate. I've looked at the output of this model many times. If the load is 240 W, the system outputs 240 W every iteration. 

I'd welcome comments.

Comment (rraustad) [inline src/EnergyPlus/HVACUnitarySystem.cc:1746]: This is a similar action to VAV terminals [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SingleDuct.cc#L2479-L2483) where oscillating flow rates are halted by choosing a specific flow.

**Comment by rraustad** [comment]
Revised the lock down methodology to only lock down operating mode if the model detects that the mode is changing (oscillating). The first figure shows the code, the second figure shows the oscillations I am trying to eliminate. Note in this example, the system is switching between heating mode and no cooling or heating every other second iteration.

This is why I picked iter > 4. Also note that before I simply locked down mode regardless of operation while now I only lock down if oscillations are detected.

maxiterfix1

maxiterfix2

**Comment by rraustad** [comment]
Interesting. DirectIndirectEvapCoolersVSAS is a problem both before and after this fix. Before in V8.4, the inlet and outlet node mass flow rate was not the same. MaxAvail was choking flow and this model did not complain.

directindirectv84

The air flow sizing for this model is troublesome. The coil air flow sizes differently than the parent.

```
Component Sizing Information, AirTerminal:SingleDuct:Uncontrolled, UNITS_GR_LF DIRECT AIR, Design Size Maximum Air Flow Rate [m3/s], 0.54498
Component Sizing Information, Branch, FURNACE_PACU_CAV:1 AIR LOOP MAIN BRANCH, Maximum Flow Rate [m3/s], 0.54498
Component Sizing Information, AirLoopHVAC, FURNACE_PACU_CAV:1, Design Supply Air Flow Rate [m3/s], 0.54498
Component Sizing Information, Controller:OutdoorAir, FURNACE_PACU_CAV:1_OA_CONTROLLER, Maximum Outdoor Air Flow Rate [m3/s], 0.54498
Component Sizing Information, COIL:COOLING:DX:VARIABLESPEED, FURNACE_PACU_CAV:1_UNITARY_PACKAGE_COOLCOIL, Design Size Rated Air Flow Rate [m3/s], 0.60325
Component Sizing Information, AirLoopHVAC:UnitaryHeatCool, FURNACE_PACU_CAV:1_UNITARY_PACKAGE, Supply Air Flow Rate [m3/s], 0.54498
Component Sizing Information, AirLoopHVAC:UnitaryHeatCool, FURNACE_PACU_CAV:1_UNITARY_PACKAGE, Supply Air Flow Rate During Heating Operation [m3/s], 0.54498
Component Sizing Information, AirLoopHVAC:UnitaryHeatCool, FURNACE_PACU_CAV:1_UNITARY_PACKAGE, Supply Air Flow Rate During Cooling Operation [m3/s], 0.54498
Component Sizing Information, AirLoopHVAC:UnitaryHeatCool, FURNACE_PACU_CAV:1_UNITARY_PACKAGE, Supply Air Flow Rate When No Cooling or Heating is Needed [m3/s], 0.54498
```

**Comment by rraustad** [comment]
FurnaceWithDXSystemRHControl looks the same as before.

furnacewithdxsystemrhcontrol_maxiterfix

**Comment by rraustad** [comment]
HP_ICSSolarCollector look the same as before.

hp_wicssolarcollector_maxiterfix

**Comment by rraustad** [comment]
HVACTemplate-5ZoneFurnaceDX looks the same.

hvactemplate-5zonefurnacedx

**Comment by rraustad** [comment]
Looking for causes of big diff's I came across a huge diff in SupermarketSubCoolersVariableSuction, as  shown in eplusout.csv.absdiff on the CI site. However, comparing V8.4 to develop I see no such huge diff.

Either the CI is comparing something different than I am, or there is an issue with calculating diff's.

supermarketsubcoolersvariablesuction_maxiterfix

**Comment by Myoldmopar** [comment]
@rraustad You have put in a huge effort in checking out this branch.  It looks like things are good to go in.  I could continue trying to spot check things, but I really don't feel like I need to.  Do you need to check anything else or need me to check anything else before this goes in?

**Comment by rraustad** [comment]
No need for spot checks. What I found, and Mike as well, was correcting this issue uncovered other deficiencies as I have pointed out here. There will be follow up issues to correct those. This was one hard nut to crack.

**Comment by Myoldmopar** [comment]
That's what I anticipated.  OK, I'm merging this in.  Thanks for the extensive effort here.

**Pull request merged**